### PR TITLE
Add functionality to the grid

### PR DESCRIPTION
**Pull request by camjc**
#### What does this PR accomplish?
GridCell:

- grow property makes the cell grow to use up all available space.
- classSuffixes can be passed to add custom classes for styling

GridHeader removed in favour of passing type='header' to a normal row.

GridRow:

- Now takes a type (defaulting to 'body') out of
  body, header, subtotal, total.
- Can turn off the horizontal border.
- Can compress the row vertically, decreasing the padding.

Removed extraneous row from the example grid.

Grid:
- Changed to border-box.
- Remove border-top when two grids are direct siblings.

#### How should this be manually tested?
- [x] Test in IE and a sane browser.

#### Any background context you want to provide?
These additions are to support required functionality for future components (immediately for the upcoming treepicker).

#### What are the relevant trello cards?
https://trello.com/c/qmGY6bxB/4756-alexandria-increase-grid-functionality-to-support-requirements-of-treepicker

#### Screenshots (if appropriate)
screen shot 2015-12-09 at 1 34 59 pm
(Future component using these features of the grid)

### CODE REVIEW

**Comment by mcslash** [comment]
IE = I(nsan)E

File: src/styles/alexandria/Grid.scss [hunk 1]
```diff
@@ -24,4 +25,8 @@
       border-right: 0;
     }
   }
+
+  + .grid-component {
```
Comment (robmunro) [inline src/styles/alexandria/Grid.scss]: :+1:

File: src/styles/alexandria/GridCell.scss [hunk 2]
```diff
@@ -1,4 +1,11 @@
+@import '../mixin';
+
 .grid-component-cell {
   flex-grow: 1;
   padding: 8px 10px;
+
+  &-grow {
+    @include ellipse;
```
Comment (robmunro) [inline src/styles/alexandria/GridCell.scss]: :+1:

**Comment by robmunro** [comment]
You sure about the horizontal borders? When do we not use them?

**Comment by robmunro** [comment]
BTW: LGTM :dancer:

File: src/components/alexandria/GridRowComponent.js [hunk 3]
```diff
@@ -1,23 +1,37 @@
...
   );
 };
+
 GridRowComponent.displayName = 'AlexandriaGridRowComponent';
 
 GridRowComponent.propTypes = {
-  verticalCellBorder: React.PropTypes.bool,
+  horizontalBorder: React.PropTypes.bool.isRequired,
+  short: React.PropTypes.bool.isRequired,
+  type: React.PropTypes.oneOf(['body', 'header', 'subtotal', 'total']).isRequired,
```
Comment (vinteo) [inline src/components/alexandria/GridRowComponent.js]: `subtotal` and `total` seem too specific

File: src/components/Main.js [hunk 4]
```diff
@@ -4,15 +4,13 @@ import _ from 'lodash';
 import Alert from 'components/alexandria/AlertComponent';
 import Breadcrumb from 'components/alexandria/BreadcrumbComponent';
 import Empty from 'components/alexandria/EmptyComponent';
+import Grid from 'components/alexandria/GridComponent';
+import GridCell from 'components/alexandria/GridCellComponent';
+import GridRow from 'components/alexandria/GridRowComponent';
 import React from 'react';
```
Comment (tchcxp) [inline src/components/Main.js]: I would put the `react` on top of all our components import.

File: src/components/alexandria/GridCellComponent.js [hunk 5]
```diff
@@ -1,14 +1,34 @@
...
       {children}
     </div>
   );
 };
+
 GridCellComponent.displayName = 'AlexandriaGridCellComponent';
 
+GridCellComponent.propTypes = {
+  classSuffixes: React.PropTypes.arrayOf(React.PropTypes.string).isRequired,
+  grow: React.PropTypes.bool.isRequired,
```
Comment (vinteo) [inline src/components/alexandria/GridCellComponent.js]: `fill` makes more sense to me than `grow`

Comment (vinteo) [inline src/components/alexandria/GridCellComponent.js]: I guess its ok since it follows the css :|

Comment (robmunro) [inline src/components/alexandria/GridCellComponent.js]: Yeah agree with @vinteo - think fill is better.

File: src/components/Main.js (hunk 4 above)
Comment (omgaz) [inline src/components/Main.js]: Yea I would agree, keep the rest alphabetical though

**Comment by camjc** [comment]
@robmunro RE: horizontal borders.
See the Top right of the picture underneath "Selected"

File: src/components/alexandria/GridCellComponent.js (hunk 5 above)
Comment (camjc) [inline src/components/alexandria/GridCellComponent.js]: @robmunro @vinteo 
The reason I don't want to call it fill is that fill sounds like its coloring in the block (like we do for the subtotal rows).

File: src/components/Main.js (hunk 4 above)
Comment (camjc) [inline src/components/Main.js]: These get automatically alphabetized by atom, I don't want to manually mess around with special casing these.

File: src/components/alexandria/GridRowComponent.js (hunk 3 above)
Comment (camjc) [inline src/components/alexandria/GridRowComponent.js]: footer and subfooter?

Comment (omgaz) [inline src/components/alexandria/GridRowComponent.js]: `footer` and `subfooter` maybe?

File: src/components/alexandria/GridCellComponent.js (hunk 5 above)
Comment (camjc) [inline src/components/alexandria/GridCellComponent.js]: Going to call it `stretch` from Vince's suggestion.

Comment (mcslash) [inline src/components/alexandria/GridCellComponent.js]: fill sounds like colour fill, but grow doesn't sound right either. stretched or something similar would be better IMO.

File: src/components/Main.js (hunk 4 above)
Comment (camjc) [inline src/components/Main.js]: Using a plugin that uses https://github.com/renke/import-sort btw.

File: src/helpers/classSuffixHelper.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import _ from 'lodash';
```
Comment (tchcxp) [inline src/helpers/classSuffixHelper.js]: We may need to add test for this helper?

Comment (camjc) [inline src/helpers/classSuffixHelper.js]: Its 100% covered by the consumers. Happy to add specific tests if its useful but I think it would be redundant.

**Comment by tchcxp** [comment]
LGTM :+1:

**Comment by etaoins** [comment]
LGTM :grimacing:

File: src/helpers/classSuffixHelper.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * Get classNames for a component.
+ *
+ * @param  {Array} classSuffixes The list of classSuffixes to add classes for.
+ * @param  {Object} suffixOptions The boolean values to add classes for when true.
+ * @param  {String} componentClass} The name of the component to add to each suffix.
```
Comment (omgaz) [inline src/helpers/classSuffixHelper.js]: one too many }

File: src/components/Main.js (hunk 4 above)
Comment (robmunro) [inline src/components/Main.js]: ?? One is external library - other is internal.

Seems like an over-optimisation if you ask me, with little benefit. This only makes sense if everyone installs that plugin - are you suggesting we do?

File: src/styles/alexandria/GridRow.scss [hunk 8]
```diff
@@ -1,7 +1,40 @@
...
+  }
+
+  &-footer {
+    font-weight: $font-weight-bold;
+  }
+
+  &-short {
+    .grid-component-cell {
+      padding-bottom: 3px;
+      padding-top: 3px;
```
Comment (omgaz) [inline src/styles/alexandria/GridRow.scss]: Does scss not have a @padding-top?

Comment (camjc) [inline src/styles/alexandria/GridRow.scss]: Nah :\

File: src/helpers/classSuffixHelper.js (hunk 7 above)
Comment (camjc) [inline src/helpers/classSuffixHelper.js]: Actually its one too few. Will fix.

**Comment by robmunro** [comment]
(shipit)?

**Comment by robmunro** [comment]
Merging!

File: src/components/Main.js (hunk 4 above)
Comment (camjc) [inline src/components/Main.js]: Its the same as a normal alphabetical sort. I'm not sure on this, you raise some good points and maybe its something we should discuss. Everything is alphabetised already so I intend to keep them that way, but I'll stop messing with other peoples ordering until we have a proper discussion.

**Pull request merged**